bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BookieShell.java

```diff
@@ -1024,41 +1022,45 @@ public int runCmd(CommandLine cmdLine) throws Exception {
 
                     Processor<Long> ledgerProcessor = new Processor<Long>() {
-                        @Override
-                        public void process(Long ledgerId, VoidCallback cb) {
-                            if (!printMeta && (bookieAddress == null)) {
-                                printLedgerMetadata(ledgerId, null, false);
-                                cb.processResult(BKException.Code.OK, null, null);
-                            } else {
-                                GenericCallback<Versioned<LedgerMetadata>> gencb = (rc, ledgerMetadata) -> {
-                                    if (rc == BKException.Code.OK) {
-                                        if ((bookieAddress == null)
-                                            || BookKeeperAdmin.areEntriesOfLedgerStoredInTheBookie(
-                                                    ledgerId, bookieAddress, ledgerMetadata.getValue())) {
-                                            /*
-                                             * the print method has to be in
-                                             * synchronized scope, otherwise
-                                             * output of printLedgerMetadata
-                                             * could interleave since this
-                                             * callback for different
-                                             * ledgers can happen in
-                                             * different threads.
-                                             */
-                                            synchronized (BookieShell.this) {
-                                                printLedgerMetadata(ledgerId, ledgerMetadata.getValue(), printMeta);
-                                            }
-                                        }
-                                    } else if (rc == BKException.Code.NoSuchLedgerExistsException) {
-                                        rc = BKException.Code.OK;
-                                    } else {
-                                        LOG.error("Unable to read the ledger: " + ledgerId + " information");
-                                    }
-                                    cb.processResult(rc, null, null);
-                                };
-                                ledgerManager.readLedgerMetadata(ledgerId, gencb);
+                            @Override
+                            public void process(Long ledgerId, VoidCallback cb) {
+                                if (!printMeta && (bookieAddress == null)) {
+                                    printLedgerMetadata(ledgerId, null, false);
+                                    cb.processResult(BKException.Code.OK, null, null);
+                                } else {
+                                    ledgerManager.readLedgerMetadata(ledgerId).whenComplete(
+                                            (metadata, exception) -> {
```

Comment:
I think the exception would be `CompletionException` here. So I am not sure how `if (exception instanceof BKException.BKNoSuchLedgerExistsException)` will  be triggered. (unless I misunderstand `whenComplete`)

see: https://github.com/apache/bookkeeper/blob/master/bookkeeper-common/src/main/java/org/apache/bookkeeper/common/concurrent/FutureEventListener.java#L35

Reply:
in this case, it would be a BKException, though when exceptions are wrapped and when they are not is not straightforward. 

If I have an future X and I call completeExceptionally on it, all handlers for that exception will get the raw exception. Any futures which have been created when installing the handlers will get the wrapped exception. 

In any case, I'm going remove all the instanceof and just pull the error codes out